File: apps/backend/app/schemas/memory.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
import uuid


class MemoryChunkCreate(BaseModel):
    chapter_id: Optional[uuid.UUID] = None
    memory_type: str = "event"
    title: Optional[str] = None
    content: str
    chapter_number: Optional[int] = None
    tags: List[str] = []
    importance_score: float = Field(default=0.5, ge=0.0, le=1.0)


class MemoryChunkOut(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    chapter_id: Optional[uuid.UUID]
    memory_type: str
    title: Optional[str]
    content: str
    chapter_number: Optional[int]
    tags: List[str]
    # 重要度 / 访问统计（时效衰减检索用）
    importance_score: float = 0.5
    access_count: int = 0
    last_accessed_at: Optional[datetime] = None
    created_at: datetime

    class Config:
        from_attributes = True


class MemoryConflictItem(BaseModel):
    """单条记忆冲突描述。"""
    conflict_type: str              # character_state / timeline / attribute / foreshadow
    severity: str                   # low / medium / high
    description: str                # 人可读的冲突说明
    chunk_ids: List[uuid.UUID]      # 涉及的 MemoryChunk ID
    chapter_refs: List[int]         # 涉及的章节编号


class MemoryConflictReport(BaseModel):
    """记忆冲突检测结果报告。"""
    project_id: uuid.UUID
    total_chunks_scanned: int
    conflicts: List[MemoryConflictItem]
    detected_at: datetime


class MemoryConflictDetectLogOut(BaseModel):
    """单次记忆冲突检测运行日志。"""
    id: uuid.UUID
    project_id: uuid.UUID
    chapter_id: Optional[uuid.UUID]
    trigger: str
    status: str
    duration_ms: int
    total_chunks_scanned: int
    conflict_count: int
    llm_call_log_id: Optional[uuid.UUID]
    error: Optional[str]
    output_payload: dict
    created_at: datetime

    class Config:
        from_attributes = True

File: apps/client/src/types/memory.ts
// ── Memory ────────────────────────────────────────────
export interface MemoryChunk {
  id: string
  project_id: string
  chapter_id?: string
  memory_type: 'event' | 'character_state' | 'foreshadow' | 'setting' | 'conflict'
  title?: string
  content: string
  chapter_number?: number
  tags: string[]
  /** AI 提取时赋值 0.0-1.0；检索时与时效衰减共同加权排序 */
  importance_score: number
  /** 被 RAG 召回的累计次数 */
  access_count: number
  /** 最近一次被召回时间（ISO 格式） */
  last_accessed_at?: string | null
  created_at: string
}

/** 单条记忆冲突描述（与后端 MemoryConflictItem 对齐） */
export interface MemoryConflictItem {
  /** character_state | timeline | attribute | foreshadow */
  conflict_type: string
  /** high | medium | low */
  severity: 'high' | 'medium' | 'low'
  description: string
  chunk_ids: string[]
  chapter_refs: number[]
}

/** 记忆冲突检测报告（与后端 detect_memory_conflicts 返回值对齐） */
export interface MemoryConflictReport {
  total_chunks_scanned: number
  conflicts: MemoryConflictItem[]
  detected_at: string
  error?: string
}

/** 单条 RAG 语义命中（与后端 RagSearchHitOut 对齐） */
export interface RagSearchHit {
  rank: number
  memory_id: string
  score?: number | null
  retrieval_source: 'semantic' | 'recency_anchor' | 'recency_fallback'
  memory_type: string
  title?: string | null
  content: string
  content_preview: string
  chapter_number?: number | null
  tags: string[]
}

/** POST /memory/rag-query 结构化响应 */
export interface RagQueryResponse {
  log_id: string
  query: string
  params: Record<string, unknown>
  status: string
  duration_ms: number
  hits: RagSearchHit[]
  memory_summary: string
  answer_hint: string
}

/** 持久化 RAG 调用日志 */
export interface RagRetrievalLog {
  id: string
  project_id: string
  chapter_id?: string | null
  source: string
  status: string
  duration_ms: number
  input_payload: Record<string, unknown>
  output_payload: Record<string, unknown>
  created_at: string
}
